base/compiler/tfuncs.jl

```diff
@@ -2037,36 +2037,36 @@ function return_type_tfunc(interp::AbstractInterpreter, argtypes::Vector{Any}, s
                     sv.restrict_abstract_call_sites = false
                     call = abstract_call(interp, ArgInfo(nothing, argtypes_vec), sv, -1)
                     sv.restrict_abstract_call_sites = old_restrict
-                    info = verbose_stmt_info(interp) ? ReturnTypeCallInfo(call.info) : false
+                    info = verbose_stmt_info(interp) ? MethodResultPure(ReturnTypeCallInfo(call.info)) : MethodResultPure()
```

Comment:
Wrapping `return_type` call into `MethodResultPure` sounds a bit weird. How about always returning `ReturnTypeCallInfo` and setups corresponding handling in the inlining algorithm?

Reply:
That's what I did originally, but I didn't really like it, because it needs exactly the same processing in inlining. `return_type` is also not a builtin, but technically a generic function that just has special const eval support. I could go either way though. If you like the other form better, I'll switch back to it.

Reply:
Okay, then let's just go with the current implementation. We can revisit this if we factor out the inlining processing better.